Graph break fixes

ani300 opened the pull request:
This PR fixes two graph breaks observed when running Llama:
- When running without TP, the default distribution strategy is `UniformModelParallelStrategy`, which causes a graph break when checking if an input is a torch.Tensor. This fixes the check to avoid the graph break.
- Recomputing the cached cosines and sines for Rotary Embeddings as part of the forward call causes a cascade of graph breaks. This PR changes the generation call and the init method for Llama to separate the cached data update from the `forward()` call, removing all graph breaks on it.

nairbv commented on scripts/inference.py:
```diff
@@ -65,7 +65,7 @@
 
 print("loading model")
-model = load_fms_llama(args.model_path)
+model = load_fms_llama(args.model_path, ntk_scaling=False)
```
I think the default is false?

nairbv commented on fms/utils/generation.py:
```diff
@@ -55,6 +55,8 @@ def generate(
     for _ in range(max_new_tokens):
         input_ids = next_input[:, -max_seq_len:]
+        # Update cacheable data in model
+        model.update_cacheable_data(input_ids, **kwargs)
```
hmm... this imposes a fixed API on every model that would use `generate`. we were trying to avoid that. There's also no enforcement, e.g., code in the GPTBigCode PR would now fail here if using this function.

ani300 replied:
I wanted to discuss this actually. I might make it so that if a model has it it runs and if it's not defined it's just ignored, but I can't think of a way to avoid the graph break other than separating this part of the function outside the main forward pass

nairbv commented:
The strategy.py fix seems straightforward enough, but I'm hesitant on some of the other changes.

It seem like this would make the rotary embeddings more complicated, and imposes more overhead on setup in the model itself as well as imposing a contract for the generate function.

I'm not sure what will happen in HF adapters -- the HF generate call wouldn't know to call our `update_cacheable_data` function.

I'm also not sure how much we really need to support `torch.compile` for the ModelParallel path. If optimizing for performance, we probably prefer TP over MP. I only added MP support because it was easier to test certain things without having to re-shard the TP checkpoints but we can probably steer people towards the TP distribution strategy for torch.compile for now.

ani300 commented on fms/distributed/strategy.py:
```diff
@@ -57,9 +57,15 @@ def __init__(self, module: nn.Module, device):
         device = self.device
-        args = [arg.to(device) if type(arg) == torch.Tensor else arg for arg in args]
+        args = [
+            arg.to(device) if isinstance(arg, torch.Tensor) else arg for arg in args
```
seems so

ani300 commented:
Re HF generate, there is a supported way (which we currently use in the closed source version) to call a function before each forward call.

Wrt making Rotary more complicated, if we pay the init cost at model init (which adds one more line to the init of Llama, it continues to work for fixed sequence lengths. Now, if you are generating, you need to call the new `update_cacheable_data`, which I was trying to make into a generic call that our models can implement and that basically takes in the same parameters as `forward` but just updates these cacheable entries.

ani300 commented:
Further updates on the PR:
- Changed behavior of RotaryEmbeddings to prevent the graph break while keeping the recomputation if needed.
- Fixed a compilation error on PT2.1 when using the kv-cache by disabling an inductor optimization pass.

nairbv commented on fms/modules/positions.py:
```diff
@@ -195,18 +196,18 @@ def compute_freqs_cis(self, device, max_seq_len=2048):
             ],
             dim=2,
-        ).view(*freqs.shape, 2, 2)
+        ).view(*freqs.size(), 2, 2)
```
lol so size vs shape fixed it

nairbv commented on scripts/inference.py:
```diff
@@ -73,6 +73,8 @@
 if args.compile:
     print("compiling model")
+    # Bug with kv-cache in PT2.1
+    torch._inductor.config.joint_graph_constant_folding = False
```
if we have an issue filed we should link to it

ani300 replied:
we don't rn, but as soon as we do I'll file it

nairbv reviewed (approved):
looks good but please add a link to the PT issue that we're working around
